crypto/dh/dh_gen.c

```diff
@@ -15,19 +15,84 @@
 #include <stdio.h>
 #include "internal/cryptlib.h"
 #include <openssl/bn.h>
+#include <openssl/dh.h>
 #include "dh_local.h"
 
+#ifndef FIPS_MODE
 static int dh_builtin_genparams(DH *ret, int prime_len, int generator,
                                 BN_GENCB *cb);
+#endif /* FIPS_MODE */
+
+int DH_generate_ffc_parameters(OPENSSL_CTX *libctx, DH *dh, int bits,
```

Comment:
Why is this public? We shouldn't be adding new public functions like this, whilst at the same time deprecating the rest of the low level functions.

Reply:
Made it private - once keygen is done we need to be able to select this method (so added a TODO(3.0) to say this.
We need to figure out what we are going to do with the dsaparam and dhparam apps that currently only call low level DH and DSA functions.